### PR TITLE
[Qtwebkit] Build with gstreamer backend

### PR DESCRIPTION
**Pull request by bagong**
MSYS2 is de facto unique in that it is easy to build qtwebkit with gstreamer backend on Windows (it is the default on Linux and has the best multimedia support of all platforms there). This enables video playback capabilities and some initial support for webaudio. Possibly multimedia support can be enhanced in this configuration while qtwebkit-5.212 stabilizes. The exposure via MSYS2 might help. The alternatives QtMultiMedia and MS-MediaFoundation do simply not work in mingw-builds.

### CODE REVIEW

**Comment by bagong** [comment]
I don't know, why the appveyor build configuration doesn't find gstreamer and plugins. They're available and were found no problem on my local builds which are totally standard msys2 builds. I ran 32- and 64-bit tests. I mean, this a monstrous build. Maybe it's just a bit much for appveyor? ;)

**Comment by lazka** [comment]
I think the problem is this line `Found PkgConfig: C:/cygwin/bin/pkg-config.exe (found version "0.29.1") ` Looks like it for some reason picks up the wrong pkg-config

**Comment by lazka** [comment]
`export PKG_CONFIG="${MINGW_PREFIX}/bin/pkg-config.exe"` might help

**Comment by bagong** [comment]
Let's see ;) This builds for me without the pkg-config variable. Let's see if it helps.

I can confirm that if properly deployed qtwebkit will have good video and partial webaudio support in this build configuration.

File: mingw-w64-qtwebkit/PKGBUILD [hunk 1]
```diff
@@ -57,6 +57,7 @@ sha256sums=('f8f901de567e11fc5659402b6b827eac75505ff9c5072d8e919aa306003f8f8a'
             'd67b006cc6d3d7b06d3a61b8649dfef1d67fb72c51f1355675d02eda17695f7b')
 
 prepare() {
+  export PKG_CONFIG="${MINGW_PREFIX}/bin/pkg-config.exe
```
Comment (lazka) [inline mingw-w64-qtwebkit/PKGBUILD]: I'd put it into build() where it's actually used. so that `--noprepare` works etc.

Comment (bagong) [inline mingw-w64-qtwebkit/PKGBUILD]: Okay

**Comment by lazka** [comment]
Hm, it finds some more from /cygwin/bin.. maybe something includes it in PATH?

**Comment by bagong** [comment]
You're right, although those are not build dependencies. But I wonder: default msys2 shields away all system paths, right. My mingw64-path is: 

```
/mingw64/bin:/usr/local/bin:/usr/bin:/bin:/c/Windows/System32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0/:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/c/Program Files (x86)/NSIS
```

The NSIS thing is only there because I added it via bashrc. So why does the appveyor build-machine add all those potentially dangerous paths?

**Comment by bagong** [comment]
My local build was done in the msys2 shell (using mingw64 and mingw32 toolchain). Bison, python2 and gperf were found in /usr/bin.

**Comment by bagong** [comment]
Now that you make me aware of it, there is even a patch that tries to prevent finding perl in cygwin, but it fails. I just added the last, gcc related patch (it's from Arch, and @annulen, the qtwebkit maintainer, made me aware of it)). The other patches are from the previous release...

**Comment by bagong** [comment]
Now the tools seem to be found correctly. At least I think that is how it worked in my local build. I must have been very lucky... So now jsmin not found...

**Comment by bagong** [comment]
I don't know how to go on with this any more. My original PR builds fine in my environment, which is not special in any way. The main difference I guess is that I don't have cygwin on my system. With each adjustment to errors I see, I move further from the original PR and potentially create problems that didn't exist before, and that I cannot fix. So unless somebody can suggest something I am close to giving up on this one. It's a pity though if webkit has no video support.

**Comment by jtanx** [comment]
Setting `DPYTHON_EXECUTABLE=/usr/bin/python2` would not be right - that's choosing the msys2/cygwin python, when you want something more like `-DPYTHON_EXECUTABLE=${MINGW_PREFIX}/python2`.

I'd also get rid of the dependency on `python2` as you already have `${MINGW_PACKAGE_PREFIX}-python2`

**Comment by bagong** [comment]
I moved python, perl, ruby, bison and gperf and addressed the executables explicitly because it seemed to me that in my local build the msys versions were being used/found, and here many of them were found in cygwin. That was purely experimental, there are subtle consequences I cannot estimate, just hope for the robustness of the build.

If that's fine with maintainers, I can try go gradually move each of those to mingw to see if it helps. It means using appveyor for a trial-error approach, if that's okay. I cannot use my local system, as it behaves differently.

**Comment by jtanx** [comment]
Some of them I would expect to be using the binaries from /usr/bin (generally anything only required by the build system), but some (like python, which is used by the package itself) should be from mingw. 

It's just weird that cmake is using anything from C:/cygwin in the first place; having to manually specify the paths shouldn't be necessary imo...

**Comment by bagong** [comment]
The mingw python2 version was outdated! That sounds weird too, hasn't python 2.7 been around for years? My local build system is latest msys2, gcc 7.1 qt5.9 etc...

**Comment by lazka** [comment]
Might be due to bin/python2 being a shell script atm. "/bin/python" might work

**Comment by annulen** [comment]
>Some of them I would expect to be using the binaries from /usr/bin (generally anything only required by the build system), but some (like python, which is used by the package itself) should be from mingw.

Python is used by build system only

**Comment by bagong** [comment]
Okay, now it found cygwin-python again. I'll try bin/python2 if this breaks again.

**Comment by jtanx** [comment]
I don't see how the cygwin python is being detected by cmake; I've tested on my own appveyor instance, and the PATH is:

```
/usr/local/bin:/usr/bin:/bin:/opt/bin:/c/Perl/site/bin:/c/Perl/bin:/c/Windows/system32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0:/c/Program Files/7-Zip:/c/Program Files/Microsoft/Web Platform Installer:/c/Tools/GitVersion:/c/Tools/PsTools:/c/Program Files/Git LFS:/c/Program Files (x86)/Subversion/bin:/c/Program Files/Microsoft SQL Server/120/Tools/Binn:/c/Program Files/Microsoft SQL Server/Client SDK/ODBC/110/Tools/Binn:/c/Program Files (x86)/Microsoft SQL Server/120/Tools/Binn:/c/Program Files/Microsoft SQL Server/120/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/120/Tools/Binn/ManagementStudio:/c/Tools/WebDriver:/c/Program Files (x86)/Microsoft SDKs/TypeScript/1.4:/c/Program Files (x86)/Microsoft Visual Studio 12.0/Common7/IDE/PrivateAssemblies:/c/Program Files (x86)/Microsoft SDKs/Azure/CLI/wbin:/c/Ruby193/bin:/c/Tools/NUnit/bin:/c/Tools/xUnit:/c/Tools/MSpec:/c/Tools/Coverity/bin:/c/Program Files (x86)/CMake/bin:/c/go/bin:/c/Program Files/Java/jdk1.8.0/bin:/c/Python27:/c/Program Files/nodejs:/c/Program Files (x86)/iojs:/c/Program Files/iojs:/c/Users/appveyor/AppData/Roaming/npm:/c/Program Files/Microsoft SQL Server/130/Tools/Binn:/c/Program Files (x86)/MSBuild/14.0/Bin:/c/Tools/NuGet:/c/Program Files (x86)/Microsoft Visual Studio 14.0/Common7/IDE/CommonExtensions/Microsoft/TestWindow:/c/Program Files/Microsoft DNX/Dnvm:/c/Program Files/Microsoft SQL Server/Client SDK/ODBC/130/Tools/Binn:/c/Program Files (x86)/Microsoft SQL Server/130/Tools/Binn:/c/Program Files (x86)/Microsoft SQL Server/130/DTS/Binn:/c/Program Files/Microsoft SQL Server/130/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/110/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/120/DTS/Binn:/c/Program Files (x86)/Apache/Maven/bin:/c/Python27/Scripts:/c/Tools/NUnit3:/c/Program Files/Mercurial:/c/Program Files/LLVM/bin:/c/Program Files/dotnet:/c/Program Files/erl8.3/bin:/c/Tools/curl/bin:/c/Program Files/Amazon/AWSCLI:/c/Program Files/Git/cmd:/c/Program Files/Git/usr/bin:/c/Program Files (x86)/Microsoft SQL Server/140/DTS/Binn:/c/Program Files (x86)/Microsoft Visual Studio 14.0/Common7/IDE/Extensions/Microsoft/SQLDB/DAC/140:/c/ProgramData/chocolatey/bin:/c/Program Files (x86)/nodejs:/c/Program Files (x86)/Yarn/bin:/c/Program Files/Microsoft Service Fabric/bin/Fabric/Fabric.Code:/c/Program Files/Microsoft SDKs/Service Fabric/Tools/ServiceFabricLocalClusterManager:/c/Users/appveyor/AppData/Roaming/npm:/c/Users/appveyor/AppData/Local/Yarn/bin:/c/Program Files/AppVeyor/BuildAgent:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl
```

Which contains no `/c/cygwin`...

**Comment by jtanx** [comment]
> Python is used by build system only

Ah, right I understand now (I think). It generates jsmin.py, and down the line it tries to use it. In that case, whatever in the build system that is setting the python path to allow `jsmin` to be found isn't working.

**Comment by jtanx** [comment]
This looks relevant? https://github.com/WebKit/webkit/commit/a0cec85c0a4112d74b8c6b36bb3783d7482c31ee

**Comment by annulen** [comment]
No,  jsmin.py is part of sources, it's located in Source/JavaScriptCore/Scripts. PYTHONPATH is set to make it available for script make-js-file-arrays.py that is located in WebCore

>This looks relevant? WebKit/webkit@a0cec85

Yes, if there is no other way I can backport that. However, existing solution with `cmake -E env` worked for all parties before.

**Comment by jtanx** [comment]
@annulen I edited my previous post; I think the python path setting code isn't right when running on mingw (note how the conditional is on WIN32, which blindly assumes that it's being built from cmd)

**Comment by annulen** [comment]
@jtanx cmake -E env is used with cmake >= 3.1, and WIN32 branch is used only as a fallback

**Comment by jtanx** [comment]
Ahh right, I missed the double negative...

**Comment by bagong** [comment]
Back to jsmin ;)

**Comment by annulen** [comment]
I've applied WebKit/webkit@a0cec85, seems to work well

**Comment by bagong** [comment]
It's passed jsmin now. Fingers crossed...

**Comment by mati865** [comment]
It ran for 1,5 hour (our AppVeyor limit) so it should be fine.

**Comment by bagong** [comment]
Cool!

**Comment by bagong** [comment]
If people think, "burdening" the qtwebkit build with several hundred MB of dependencies might be too much, maybe it would be an option to offer two packages, one "basic" and one "gst"? The build is so large (and maybe a bit fragile too, although I didn't run into many errors myself), that having binaries provided from msys2, built in the current environment, would be a really great service. Building qtwebkit takes almost as long as building all of Qt proper ;)

**Comment by bagong** [comment]
I am happy to retest with updated gstreamer once that PR gets committed...

**Comment by rlatapy-gpsw** [comment]
Hi @bagong 
I'm new at msys build system, I'm trying to build QtWebkit with gstreamer using your mingw package, and obviously without success..
If I well understand, I just have to go into the mingw-w64-qtwebkit and call MINGW_INSTALLS=mingw64 makepkg-mingw -sLf ?

[mingw-w64-qtwebkit-5.212.0alpha2-2-x86_64-build.txt](https://github.com/Alexpux/MINGW-packages/files/1169326/mingw-w64-qtwebkit-5.212.0alpha2-2-x86_64-build.txt)
[mingw-w64-qtwebkit-5.212.0alpha2-2-x86_64-prepare.txt](https://github.com/Alexpux/MINGW-packages/files/1169327/mingw-w64-qtwebkit-5.212.0alpha2-2-x86_64-prepare.txt)

**Comment by bagong** [comment]
@rlatapy-gpsw, I am not sure what causes "could not open FormState.h". Also I can't see what happened before "prepare"

This seems to be the error:
`````
[35m[1mScanning dependencies of target qmlwebkitexperimentalplugin-files[0m
Could not open C:/msys64/home/kolor/mingw-w64-qtwebkit/src/build-x86_64-w64-mingw32-debug/DerivedSources/ForwardingHeaders/WebCore/FormState.h. at C:/msys64/home/kolor/mingw-w64-qtwebkit/src/qtwebkit-5.212.0-alpha2/Source/WebKit2/Scripts/generate-forwarding-headers.pl line 120, <EXISTING_HEADER> line 1.
make[2]: *** [Source/WebKit2/CMakeFiles/WebKit2-forwarding-headers.dir/build.make:57: Source/WebKit2/CMakeFiles/WebKit2-forwarding-headers] Error 13
[  0%] [34m[1mGenerating ../../DerivedSources/ForwardingHeaders/JavaScriptCore/Scripts/make-js-file-arrays.py[0m
[  0%] [32mBuilding CXX object Source/WTF/wtf/CMakeFiles/WTF.dir/Atomics.cpp.obj[0m
make[1]: *** [CMakeFiles/Makefile2:2198: Source/WebKit2/CMakeFiles/WebKit2-forwarding-headers.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
````

If you're willing to follow "guesses":

- use the msys2 shell (not the mingw64 one)
- run the msys2 shell as administrator. If you don't do that, you'll get errors concerning symbolic links at an early stage of prepare, maybe they cause the later failure


Let me know if that helps...

**Comment by rlatapy-gpsw** [comment]
Using msys2 shell instead of mingw64 lead to same result, I don't try running as admin yet.
Btw, I disable webkit2 in PKGBUILD and build works. But I cannot play video in MiniBrowser or QtTestBrowser...
image

**Comment by bagong** [comment]
Likely that is because you haven't deployed gstreamer yet. The plugins and some additional libraries need to be made available to the binary using qtwebkit. A rule of thumb is: (copy from the mingw64-folder to the folder where your executable resides):

- the entire folder lib/gstreamer-1.0 (keep the lib folder)
- all dlls starting with bin/ libgst*, libtheora*, libtiff*, libvorbis*, libvpx*, libwavpack*, libwebp (no need to copy to a bin folder, side by side with the exec is fine)
- two openssh libraries (for ssl-support): libeay32.dll, ssleay32.dll

I haven't tested Minibrowser, but QtTestbrowser and our app gives me wide ranging video support. (and webkit2 also builds)

**Comment by rlatapy-gpsw** [comment]
I forgot the gstreamer lib folder...
Msys shell running as admin seems to solve building issue with webkit2
So thank you everything works

**Comment by bagong** [comment]
Since the gstreamer update QtTestBrowser and QtMinibrowser fail to start returning a segmentation fault. @annulen , should one try to get to the ground of this?

**Comment by annulen** [comment]
If you need help with segmentation fault, you should at least provide backtrace

**Pull request closed without merging**